### PR TITLE
Rest api feature

### PR DESCRIPTION
**Pull request by ubbabeck**
Added method `getCFilterRest` to `query.go` that queries bitcoin node using https request.

In order to enable this I have also added a new field `RestPeers [] string[]` to the neutrino config in neutrino.go so that the client can spesify who they wish to connect to. 

Functionality to use a tor proxy through `127.0.0.1:9050` is also enabled by adding `TorProxy bool` to the neutrino config in `neutrino.go`

What I in the first case would like the reviewers to focus on is any obious mistakes, bugs and the general design of the solution. Second how I am doing the peer selection in `neutrino.go` and how to make it dynamic within `GetCFilter`. Even though best results where acheived when querying one host at a time it would be nice to have a fallback to an alternaltive host, if given or p2p.

Results:
Querying blocks 200 blocks CFilter with Rest API  took 1m12.992833146s.
Querying blocks 200 blocks CFilter with bitcoin P2P took 3m39.44101219s.

### CODE REVIEW

File: neutrino.go [hunk 1]
```diff
@@ -945,6 +966,43 @@ func NewChainService(cfg Config) (*ChainService, error) {
...
+		s.client, err = NewHTTPClient(cfg.TorProxy)
+		if err != nil {
+			log.Debugf("error setting up HTTPClient: %w", err)
+		}
+		// Iterating thought restpeer defined in the config
+		// and checking if they are reachable.
+		for _, restAddr := range restPeers {
+			u, err := url.Parse(restAddr)
+			if err != nil {
+				log.Debugf("error: %w", err)
```
Comment (roeierez) [inline neutrino.go]: shouldn't we need to break here?

File: neutrino.go [hunk 2]
```diff
@@ -945,6 +966,43 @@ func NewChainService(cfg Config) (*ChainService, error) {
...
+			}
+
+			_, err = s.nameResolver(u.Host)
```
Comment (roeierez) [inline neutrino.go]: Why do we need this step? We anyway query the url later

File: neutrino.go [hunk 3]
```diff
@@ -626,6 +636,11 @@ type peerSubscription struct {
 	cancel <-chan struct{}
 }
 
+// HTTPClient holds a HTTP client in order to query using rest api if enabled.
+type HTTPClient struct {
```
Comment (roeierez) [inline neutrino.go]: Why do we need to wrap http.Client inside this struct? Isn't it simpler to just hold the http.Client directly?

File: query.go [hunk 4]
```diff
@@ -892,36 +925,79 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
...
+					return
+				}
+				hash := blockHeaders.BlockHash()
+
+				filter, err := s.getCFilterRest(hash, RestHostIndex)
+				if err != nil {
+					fmt.Errorf("error: %w", err)
+					return
+				}
+				quit := make(chan struct{})
```
Comment (roeierez) [inline query.go]: we should create this channel once outside of the loop

**Review by roeierez — CHANGES_REQUESTED** [review]
@ubbabeck I did initial review, dropped some comments.

File: neutrino.go (hunk 3 above)
Comment (ubbabeck) [reply neutrino.go]: totally, I have now removed the struct and added just the client.

File: neutrino.go (hunk 2 above)
Comment (ubbabeck) [reply neutrino.go]: No, it has now been removed.

File: query.go (hunk 4 above)
Comment (ubbabeck) [reply query.go]: Done

File: neutrino.go [hunk 5]
```diff
@@ -945,6 +957,39 @@ func NewChainService(cfg Config) (*ChainService, error) {
...
+		// Iterating thought restpeer defined in the config
+		// and checking if they are reachable.
+		for _, restAddr := range restPeers {
+			_, err := url.Parse(restAddr)
+			if err != nil {
+				log.Debugf("error unable to parse address: %w", err)
+				break
+			}
+
+			res, err := s.client.Get(restAddr)
```
Comment (yaslama) [inline neutrino.go]: The server is not supposed to answer to GET queries here, even if one specific implementation returns OK, it's wrong to check if a server is ok at the init stage

File: neutrino.go [hunk 6]
```diff
@@ -985,6 +1030,17 @@ func NewChainService(cfg Config) (*ChainService, error) {
 	return &s, nil
 }
 
+// NewHTTPClient of type HTTPClient.
+func NewHTTPClient() (*http.Client, error) {
```
Comment (yaslama) [inline neutrino.go]: I don't see the need of this function which is a oneliner

File: neutrino.go [hunk 7]
```diff
@@ -985,6 +1030,17 @@ func NewChainService(cfg Config) (*ChainService, error) {
...
+	// We'll retun a reguar HTTP client.
+	return &http.Client{Timeout: 10 * time.Second}, nil
+}
+
+// Function queries the HTTPClient and with the requested url.
```
Comment (yaslama) [inline neutrino.go]: Why a function instead of one line of code?

File: neutrino.go [hunk 8]
```diff
@@ -671,6 +677,12 @@ type ChainService struct { // nolint:maligned
 	// will notify each time a new peer is connected.
 	peerSubscribers []*peerSubscription
 
+	// restPeers is a slice of peers that suppoorts the rest API
+	restPeers []string
+
+	// http clients
+	client *http.Client
```
Comment (yaslama) [inline neutrino.go]: Don't need to store anything in this variable. It's a simple almost empty structure: &http.Client{Timeout: 10 * time.Second}

File: neutrino.go [hunk 9]
```diff
@@ -671,6 +677,12 @@ type ChainService struct { // nolint:maligned
 	// will notify each time a new peer is connected.
 	peerSubscribers []*peerSubscription
 
+	// restPeers is a slice of peers that suppoorts the rest API
+	restPeers []string
```
Comment (yaslama) [inline neutrino.go:681]: This variable is not needed if we don't check at init time

File: query.go [hunk 10]
```diff
@@ -58,6 +61,10 @@ var (
 	// `getdata` and other similar messages.
 	QueryEncoding = wire.WitnessEncoding
 
+	// RestHostIndex specifies the current host to query using if the
+	// rest API is enabled.
+	RestHostIndex = 0
```
Comment (yaslama) [inline query.go]: Why RestHostIndex and not restHostIndex?

File: query.go [hunk 11]
```diff
@@ -892,36 +925,76 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
...
-			// the query early.
-			func(_ *ServerPeer, resp wire.Message, quit chan<- struct{}) {
-				s.handleCFiltersResponse(query, resp, quit)
-			},
-			query.options...,
-		)
+	// First attempting to query the rest API and if optimistic batching is
+	// we'll query for blocks
+	// if node does not support the rest API, we'll query using bitcoin's p2p
+	if len(s.restPeers) != 0 {
```
Comment (yaslama) [inline query.go]: I think it's better to have one goroutine (go func() {}) and the if len(s.restPeers) !=0 inside it.

**Review by yaslama — CHANGES_REQUESTED** [review]

File: query.go [hunk 12]
```diff
@@ -892,27 +926,54 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
...
-			func(_ *ServerPeer, resp wire.Message, quit chan<- struct{}) {
-				s.handleCFiltersResponse(query, resp, quit)
-			},
-			query.options...,
-		)
-
+		if len(s.restPeers) != 0 {
+			quit := make(chan struct{})
+			// We'll need a http client in order to query the host
+			client := &http.Client{Timeout: 10 * time.Second}
```
Comment (yaslama) [inline query.go]: It's perhaps a good idea to use the existing variable QueryTimeout instead of the hard-coded values

File: neutrino.go [hunk 13]
```diff
@@ -945,6 +953,22 @@ func NewChainService(cfg Config) (*ChainService, error) {
 		permanentPeers = cfg.AddPeers
 	}
 
+	// Adding rest peers to chainservice if spesified in config
+
+	restPeers := cfg.RestPeers
```
Comment (yaslama) [inline neutrino.go]: This variable is really needed? (cfg.RestPeers can be directly used in the loop)

File: neutrino.go [hunk 14]
```diff
@@ -945,6 +953,22 @@ func NewChainService(cfg Config) (*ChainService, error) {
...
+	if len(restPeers) != 0 {
```
Comment (yaslama) [inline neutrino.go]: Enclosing a for by `if len()!=0` is not redundant?

File: neutrino.go [hunk 15]
```diff
@@ -945,6 +953,22 @@ func NewChainService(cfg Config) (*ChainService, error) {
...
+		// Iterating thought restpeer defined in the config
+		// and checking if they are reachable.
```
Comment (yaslama) [inline neutrino.go]: We are not checking that it's reachable, only that the url is ok

File: neutrino.go [hunk 16]
```diff
@@ -945,6 +953,22 @@ func NewChainService(cfg Config) (*ChainService, error) {
...
+		for _, restAddr := range restPeers {
+			_, err := url.Parse(restAddr)
+			if err != nil {
+				log.Debugf("error unable to parse address: %w", err)
+				break
```
Comment (yaslama) [inline neutrino.go]: Are you sure that continue is not better? With break, if the first url is bad, all the elements of the array restPeers are lost.

**Review by yaslama — CHANGES_REQUESTED** [review]

File: query.go [hunk 17]
```diff
@@ -892,27 +926,54 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
 		return nil, err
 	}
 
+	// We will first check if the node supports rest API
+	// and try to get the filter from there
+
 	// With all the necessary items retrieved, we'll launch our concurrent
 	// query to the set of connected peers.
-	log.Debugf("Fetching filters for heights=[%v, %v], stophash=%v",
+	log.Debugf("Fetching filters for heights=[%v, %v], stophash=%v\n",
```
Comment (yaslama) [inline query.go]: Are you sure that "\n" is needed?

File: neutrino.go (hunk 6 above)
Comment (ubbabeck) [reply neutrino.go]: Removed

File: neutrino.go (hunk 7 above)
Comment (ubbabeck) [reply neutrino.go]: Added a oneliner to query.go instead

File: query.go (hunk 10 above)
Comment (ubbabeck) [reply query.go]: Edited

File: query_rest.go [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+)
+
+func (s *ChainService) queryRestPeers(
+	blockHash chainhash.Hash, query *cfiltersQuery) {
```
Comment (ubbabeck) [inline query_rest.go]: Unused param `blockHash`, if we have everything we need in query I can remove it.

File: query_rest.go [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+	for i, p := range s.restPeers {
+		if p.failures == 0 || time.Since(p.lastFailure) > 10*time.Second {
+			validPeers = append(validPeers, i)
+		}
+	}
+	if len(validPeers) == 0 {
+		log.Errorf("queryRestPeers - No valid rest peer")
+		return
+	}
+	restPeerIndex := validPeers[rand.Intn(len(validPeers))]
```
Comment (ubbabeck) [inline query_rest.go:27]: Guess do not need to care about this message.
> Use of weak random number generator (math/rand instead of crypto/rand) (gosec)

**Comment by ubbabeck** [comment]
Merged in the production branch https://github.com/breez/neutrino/tree/breez-client

**Pull request closed without merging**